sync/statesync/code_queue.go

```diff
@@ -0,0 +1,260 @@
+// Returns errAddCodeAfterFinalize after a clean finalize and errFailedToAddCodeHashesToQueue on early quit.
+func (q *CodeQueue) AddCode(codeHashes []common.Hash) error {
+	// If the queue has been closed, reject new work.
+	switch q.closed.get() {
+	case closeStateFinalized:
+		return errAddCodeAfterFinalize
+	case closeStateQuit:
+		return errFailedToAddCodeHashesToQueue
+	}
+	return q.enqueue(codeHashes)
+}
+
+// init enqueues any persisted code markers found on disk.
+func (q *CodeQueue) init() error {
+	// Recover any persisted code markers and enqueue them.
+	// Note: dbCodeHashes are already present as "to-fetch" markers. addCode will
+	// re-persist them, which is a trivial redundancy that happens only on resume
+	// (e.g., after restart). We accept this to keep the code simple.
+	dbCodeHashes, err := recoverUnfetchedCodeHashes(q.db)
+	if err != nil {
+		return fmt.Errorf("unable to recover previous sync state: %w", err)
+	}
+	if err := q.enqueue(dbCodeHashes); err != nil {
+		return fmt.Errorf("unable to resume previous sync: %w", err)
+	}
+
+	return nil
+}
+
+// enqueue persists markers for the provided hashes and pushes them onto the output channel.
+// No deduplication is performed. The consumer is responsible for dedupe and retry.
+func (q *CodeQueue) enqueue(codeHashes []common.Hash) error {
+	if len(codeHashes) == 0 {
+		return nil
+	}
+
+	batch := q.db.NewBatch()
+	// Persist all input hashes as to-fetch markers. Consumer will dedupe and skip
+	// already-present code. Persisting all enables consumer-side retry.
+	// Note: markers are keyed by code hash, so repeated persists overwrite the same
+	// key rather than growing DB usage. The consumer deletes the marker after
+	// fulfilling the request (or when it detects code is already present).
+	selected := make([]common.Hash, 0, len(codeHashes))
+	for _, codeHash := range codeHashes {
+		customrawdb.AddCodeToFetch(batch, codeHash)
+		selected = append(selected, codeHash)
+	}
+
+	if err := batch.Write(); err != nil {
+		return fmt.Errorf("failed to write batch of code to fetch markers due to: %w", err)
+	}
+
+	q.enqueueWG.Add(1)
+	defer q.enqueueWG.Done()
+	for _, hash := range selected {
+		select {
+		case q.out <- hash:
+		case <-q.quit:
+			return errFailedToAddCodeHashesToQueue
+		}
+	}
+	return nil
+}
+
+// Finalize implements [syncpkg.CodeRequestQueue] by signaling no further code hashes will be added.
+// Waits for in-flight enqueues to complete, then closes the output channel.
+// If the queue was already closed due to early quit, returns errFailedToFinalizeCodeQueue.
+func (q *CodeQueue) Finalize() error {
+	q.enqueueWG.Wait()
+	// If already closed due to quit, report early-exit error.
+	if q.closed.get() == closeStateQuit {
+		return errFailedToFinalizeCodeQueue
+	}
+	// Mark as finalized and close the channel (no-op if already finalized).
+	q.closed.markFinalizedAndClose(q.out)
+	return nil
```

Comment:
This can race. `closed.markFinalizedAndClose()` needs to propagate the boolean from `transition()` and, if false, then you can return the error here.